.github/workflows/build.yml

```diff
@@ -1,71 +1,113 @@
-
+  build-setup:
+    runs-on: 'ubuntu-latest'
     permissions:
       contents: 'read'
       id-token: 'write'
+    outputs:
+      gcloud_access_token: '${{ steps.auth.outputs.access_token }}'
+    steps:
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@8b122486cedac8393e77aa9734c3528886e4a1a8 # ratchet:docker/setup-qemu-action@v2
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@dc7b9719a96d48369863986a06765841d7ea23f6 # ratchet:docker/setup-buildx-action@v2
 
+  build:
     runs-on: 'ubuntu-latest'
-
+    needs: 'build-setup'
+    permissions:
+      contents: 'read'
+      id-token: 'write'
     steps:
+      - name: 'Configure environment variables'
+        if: ${{ inputs.env-context != '{}' }}
+        run: |-
+          jq -r <<< '${{ inputs.env-context }}' | jq -r "to_entries | map(\"\(.key)=\(.value|tostring)\") | .[]" >> $GITHUB_ENV
       - name: 'Checkout'
         uses: 'actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b' # ratchet:actions/checkout@v3
-
-      - name: 'Setup Go'
-        uses: 'actions/setup-go@fcdc43634adb5f7ae75a9d7a9b9361790f7293e2' # ratchet:actions/setup-go@v3
+      - id: 'auth'
+        name: 'Authenticate to Google Cloud'
+        uses: 'google-github-actions/auth@ceee102ec2387dd9e844e01b530ccd4ec87ce955' # ratchet:google-github-actions/auth@v0
         with:
-          go-version: '1.19'
-
-      - name: 'Cache dependencies'
-        uses: 'actions/cache@48af2dc4a9e8278b89d7fa154b955c30c6aaab09' # ratchet:actions/cache@v3
+          workload_identity_provider: '${{ env.wif_provider }}'
+          service_account: '${{ env.wif_service_account }}'
+          token_format: 'access_token'
+      - name: 'Authenticate to Artifact Registry'
+        uses: 'docker/login-action@49ed152c8eca782a232dede0303416e8f356c37b' # ratchet:docker/login-action@v2
         with:
-          path: |-
-            ~/go/pkg/mod
-          key: ${{ runner.os }}-go-test-${{ hashFiles('**/go.sum') }}
-          restore-keys: |-
-            ${{ runner.os }}-go-
-            ${{ runner.os }}-go-test-
-
+          username: 'oauth2accesstoken'
+          password: '${{ steps.auth.outputs.access_token }}'
+          registry: '${{ env.docker_registry }}'
+      - name: 'Build the main server container and push to the registry'
+        uses: 'docker/build-push-action@c84f38281176d4c9cdb1626ffafcd6b3911b5d94' # ratchet:docker/build-push-action@v3
+        with:
+          push: true
+          tags: '${{ env.docker_repo }}/lumberjack-server:${{ env.docker_tag }}'
+          file: 'scripts/server.dockerfile'
+      - id: 'set_output'
+        run: |-
+          echo "::set-output name=image_server::${{ env.docker_repo }}/lumberjack-server:${{ env.docker_tag }}"
+  build-integration-test-servers:
+    runs-on: 'ubuntu-latest'
+    permissions:
+      contents: 'read'
+      id-token: 'write'
+    needs: 'build-setup'
+    strategy:
+      fail-fast: true
+      matrix:
```

Comment:
I think I'm missing some context. Don't we have the WIF actions shared in abcxyz/pkg? 

The original integration workflow was 60+ lines but now it's 160+ lines. Is it because we relocate the logic from script to the workflow? What's the tradeoff?

Reply:
@sethvargo That's fair. Personally I find it easier to reason on but I'm sure that comes from having written in. If I break apart the matrix strategy it will be a lot of duplicated code I'd probably favor a reusable workflow over just copy-paste. I have a personal aversion to large files, especially when its YAML.

Reply:
@yolocs - There was a lot of things done in the shell script that needed to be replicated in the github workflow.   

I'm not sure I follow your WIF comment, can you elaborate?

Reply:
@yolocs - There was a lot of things done in the shell script that needed to be replicated in the github workflow.

I'm not sure I follow your WIF comment, can you elaborate?